Fix no_default/no_result comparison against custom types.

rolando opened the pull request:
When the intermediate output of reduction is a type with custom
comparison, the code breaks due to the `obj == <string>` comparison (for
instance, a ``csr_matrix``).

mrocklin commented:
The reason we've been using value equality is because sometimes the `--no-default--` string comes through after having been serialized.

```python
In [1]: a = 'abc'

In [2]: b = 'abc'

In [3]: a is b
Out[3]: True

In [4]: a == b
Out[4]: True

In [5]: import pickle

In [6]: c = pickle.loads(pickle.dumps(a))

In [7]: a is c
Out[7]: False

In [8]: a == c
Out[8]: True
```

An alternative approach would be to first check the type and then check value equality.

mrocklin commented:
Can you include a test to make sure that your fix doesn't get undone in the future?

rolando commented:
@mrocklin Oh, that's something I didn't expect. Unless I'm missing something, it seems it has to do with the `pickle` protocol:

```
$ python
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:24:55)
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> a = 'x'
>>> b = pickle.loads(pickle.dumps(a, protocol=-1))
>>> a is b
True
```

```
Python 2.7.12 |Continuum Analytics, Inc.| (default, Jun 29 2016, 11:09:23)
[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import pickle
>>> a = 'x'
>>> b = pickle.loads(pickle.dumps(a, protocol=-1))
>>> a is b
True
```

Is there a particular reason to use `protocol=0`? Or is the idea is to support non-pickle serialization methods? In that case, it makes sense to check the type for a safe string equality comparison.

rolando commented:
@mrocklin Updated the PR to focus on the ``no_result`` sentinel and added a little helper ``eq_strict`` to do the type check: https://github.com/dask/dask/pull/1442/files#diff-64b17ee9155433cad463c80a971a3223R759

Maybe all sentinel checks (i.e.: ``no_default``) should use ``eq_strict`` but that's out of the scope of this PR.

mrocklin commented:
Looks good.  Personally I would probably use `type(p) is str and p == no_result` in order to avoid the extra function (people reading the code now have to go and look up what eq_strict is supposed to do) but that's pretty subjective.  I can see reasons why eq_strict is nice as well.

Regardless, it'd be nice to have a test for dask.bag that exercises this issue.  Otherwise I can imagine someone coming by in the future and reverting this change without having a test to tell them that they're causing a regression.

rolando commented:
Fair enough. I've added a test replicating the issue using `scipy.sparse.csr_matrix`. This because:

* There are already some tests that use ``scipy``.
* And mocking the behavior with a custom class required unnecessary boilerplate.

**Edit:** I realized that ``csr_matrix`` may change its behavior in the future so I added an additional test with a custom class.

mrocklin commented:
Thanks!  Merging.
